wandb/sdk/launch/docker.py

```diff
@@ -64,139 +51,151 @@ def generate_docker_image(
+    try:
+        docker.run(["docker", "pull", docker_image])
+    except DockerError as e:
+        raise LaunchError("Docker server returned error: {}".format(e))
+
 
-    Raises:
-        LaunchError: if there is an issue communicating with the docker client
+def build_docker_image_if_needed(
+    launch_project: _project_spec.LaunchProject, api: Api, copy_code: bool
+) -> str:
+    """
+    Build a docker image containing the project in `work_dir`, using the base image.
+    param launch_project: LaunchProject class instance
+    :param api: instance of wandb.apis.internal Api
+    :param copy_code: boolean indicating if code should be copied into the docker container
     """
-    import docker
 
-    image_name = "wandb_launch_{}".format(launch_project.run_id)
     image_uri = _get_docker_image_uri(
-        name=image_name, work_dir=launch_project.project_dir
+        name=launch_project.image_name, work_dir=launch_project.project_dir
     )
+    launch_project.docker_image = image_uri
+    if docker_image_exists(image_uri):
+        wandb.termlog("Using existing image: {}".format(image_uri))
+        return image_uri
+    container_inspect = docker_image_inspect(launch_project.base_image)
     copy_code_line = ""
-    workdir_line = ""
-    copy_config_line = ""
-    workdir = os.path.join("/home/", getpass.getuser())
-    if launch_project.override_config:
-        copy_config_line = "COPY {}/{} {}\n".format(
-            _PROJECT_TAR_ARCHIVE_NAME, _project_spec.DEFAULT_CONFIG_PATH, workdir
-        )
+    requirements_line = ""
+    workdir = container_inspect["ContainerConfig"].get("WorkingDir", "/")
+    # TODO: we currently assume the home directory holds the pip cache
+    homedir = workdir
+    # for custom base_images we attempt to introspect the homedir
+    for env in container_inspect["ContainerConfig"]["Env"]:
+        if env.startswith("HOME="):
+            homedir = env.split("=", 1)[1]
     if copy_code:
-        copy_code_line = "COPY {}/ {}\n".format(_PROJECT_TAR_ARCHIVE_NAME, workdir)
-        workdir_line = "WORKDIR {}\n".format(workdir)
+        copy_code_line = "COPY ./src/ {}\n".format(workdir)
+        if docker.is_buildx_installed():
+            requirements_line = (
+                "RUN --mount=type=cache,target={}/.cache,uid={},gid=0 ".format(
+                    homedir, launch_project.docker_user_id
+                )
+            )
+        else:
+            wandb.termwarn(
+                "Docker BuildX is not installed, for faster builds upgrade docker: https://github.com/docker/buildx#installing"
+            )
+            requirements_line = "RUN WANDB_DISABLE_CACHE=true "
+        shutil.copy(
+            os.path.join(os.path.dirname(__file__), "templates", "_wandb_bootstrap.py"),
+            os.path.join(launch_project.project_dir),
+        )
+        # TODO: make this configurable or change the default behavior...
+        requirements_line += _parse_existing_requirements(launch_project)
```

Comment:
By default we're only installing the pinned versions of libraries specified in the original git repositories `requirements.txt`.  There will likely be cases where either:

a. we don't have a source `requirements.txt`
b. the user wants to install all libraries from `requirements.froxen.txt`

Hoping we could add a flag to the launch config that controlled this behavior.